demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c

```diff
@@ -0,0 +1,1354 @@
+    int returnStatus = EXIT_SUCCESS;
+
+    /* coreMQTT library return status. */
+    MQTTStatus_t mqttStatus = MQTTSuccess;
+
+    /* OTA library return status. */
+    OtaErr_t otaRet = OtaErrNone;
+
+    /* OTA Agent state returned from calling OTA_GetAgentState.*/
+    OtaState_t state = OtaAgentStateStopped;
+
+    /* OTA event message used for sending event to OTA Agent.*/
+    OtaEventMsg_t eventMsg = { 0 };
+
+    /* OTA library packet statistics per job.*/
+    OtaAgentStatistics_t otaStatistics = { 0 };
+
+    /* OTA Agent thread handle.*/
+    pthread_t threadHandle;
+
+    /* OTA interface context required for library interface functions.*/
+    OtaInterfaces_t otaInterfaces;
+
+    /* Set OTA Library interfaces.*/
+    setOtaInterfaces( &otaInterfaces );
+
+    /****************************** Init OTA Library. ******************************/
+
+    if( returnStatus == EXIT_SUCCESS )
+    {
+        if( ( otaRet = OTA_Init( &otaBuffer,
+                                 &otaInterfaces,
+                                 ( const uint8_t * ) ( CLIENT_IDENTIFIER ),
+                                 otaAppCallback ) ) != OtaErrNone )
+        {
+            LogError( ( "Failed to initialize OTA Agent, exiting = %u.",
+                        otaRet ) );
+
+            returnStatus = EXIT_FAILURE;
+        }
+    }
+
+    /****************************** Create OTA Task. ******************************/
+
+    if( returnStatus == EXIT_SUCCESS )
+    {
+        if( pthread_create( &threadHandle, NULL, otaAgentTask, NULL ) != 0 )
+        {
+            LogError( ( "Failed to create OTA thread: "
+                        ",errno=%s",
+                        strerror( errno ) ) );
+
+            returnStatus = EXIT_FAILURE;
+        }
+    }
+
+    /****************************** OTA Demo loop. ******************************/
+
+    if( returnStatus == EXIT_SUCCESS )
+    {
+        /* Wait till OTA library is stopped, output statistics for currently running
+         * OTA job */
+        while( ( ( state = OTA_GetState() ) != OtaAgentStateStopped ) )
+        {
+            if( mqttSessionEstablished != true )
+            {
+                /* Connect to MQTT broker and create MQTT connection. */
+                returnStatus = establishConnection();
+
+                if( returnStatus == EXIT_SUCCESS )
+                {
+                    mqttSessionEstablished = true;
+
+                    /* Check if OTA process was suspended and resume if required. */
+                    if( state == OtaAgentStateSuspended )
+                    {
+                        /* Resume OTA operations. */
+                        OTA_Resume();
+                    }
+                    else
+                    {
+                        /* Send start event to OTA Agent.*/
+                        eventMsg.eventId = OtaAgentEventStart;
+                        OTA_SignalEvent( &eventMsg );
+                    }
+                }
+            }
```

Comment:
Should the logic of establishing MQTT connection be present outside of the while loop (as the MQTT connection establishment should be required only once, right?)